### PR TITLE
[ramda] allow arrays to be non-readonly

### PR DESCRIPTION
**Pull request by millsp**
Ramda itself does not enforce arrays to be readonly. Instead, it lets programmers to choose whether they want to have immutable data structures or not. This can rather be achieved by using linters, and should not be enforced by default. 

In addition to the reasons above, I also found out that it makes Ramda harder to use when dealing/combining functions from external libraries that do not return `readonly` types.

I also updated ts-toolbelt's version.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` should be present and it shouldnt have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [ ] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by pirix-gh** [comment]
The PR is less scary than it looks like. It's a simple replacement of `ReadonlyArray` to `Array` or `[ ]`

**Comment by orta** [comment]
Yep, does looks scary - but your reasoning seems fine and it all looks good to me :+1:

**Comment by pirix-gh** [comment]
Oh thank you so much @orta :tada: 

I work a lot with ramda. And I also work a lot with types. My dream is to reach 90% type-compliance for Ramda. I plan to achieve this by integrating Ramda with my ts-toolbelt (computes complex types).
So... I would like to ask you if you if it would be possible for me to join the board to manage Ramda (not at full time though). So that I can work more easily on Ramda, and help with merging PRs too.

**Comment by pirix-gh** [comment]
Thanks @Christof!!

**Comment by mankittens** [comment]
@pirix-gh 

Hey, I think this change broke at least one type definition. I'm noticing that flatten no longer takes a double-array and turns it into a single array. Not sure why it behaves this what, but it's what I'm seeing.

Screen Shot 2019-09-19 at 5 10 23 PM

This update created a total of 20 errors in my project.

Screen Shot 2019-09-19 at 5 10 41 PM

**Comment by pirix-gh** [comment]
I'll fix this soon @mankittens. Thanks for reporting this issue

**Comment by pirix-gh** [comment]
Hey @mankittens. This has been fixed in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/38503. It also improves the previous types by handling unlimited array nesting.

**Comment by mankittens** [comment]
Thanks for jumping on the issue! I'll report back if I find any issues with the update. :)

File: types/ramda/index.d.ts [hunk 1]
```diff
@@ -1678,7 +1678,7 @@ declare namespace R {
          * Returns a list containing the names of all the enumerable own
          * properties of the supplied object.
          */
-        keys<T extends object>(x: T): Array<keyof T>;
+        keys<T extends object>(x: T): keyof T[];
```
[the remark this answers is not in the document]
Comment (rdhelms) [reply types/ramda/index.d.ts]: Never mind!
Just saw https://github.com/DefinitelyTyped/DefinitelyTyped/pull/38467 which will fix this
👍

**Comment by pirix-gh** [comment]
@zauberpony I get your point! This is on me, my bad. I'll fix this right away. I assumed that `readonly` would apply to `array`, while it was the other way around. You were right, this came from an edge case scenario and I got confused on my late work hours. Sorry! :see_no_evil:

**Comment by pirix-gh** [comment]
@zauberpony the fix is out https://github.com/DefinitelyTyped/DefinitelyTyped/pull/38503, let me know if anything is wrong.

**Comment by zauberpony** [comment]
Thank you very much for your quick response and changes :)

I just upgraded to `@types/ramda@0.26.24` and tsc is still happy, so it looks good to me.

**Pull request merged**